### PR TITLE
Store: Stripe Connect: Wire the disconnect link through redux

### PR DESCRIPTION
**Pull request by allendav**
See #17002

To test:
* Follow the steps in #19115 to setup your environment
* **Edit: Important: Use the fix/1203-explicit-status-stripe-rest-api-errors branch when cloning woocommerce-services**
* After you've used the test bench to connect the site, navigate to http://calypso.localhost:3000/store/settings/{DOMAIN}
* Click on the Manage link for Stripe
* Wait for the placeholder dialog to be replaced by a dialog showing you're connected, and which includes a Disconnect link
* Click the Disconnect link
* Verify you are disconnected and dropped on the key based flow dialog
* Force refresh the page and enter Manage again
* Verify you are dropped on the key based flow

* Run `npm run test-client client/extensions/woocommerce/state/` and verify all tests pass

### CODE REVIEW

**Comment by allendav** [comment]
Now with unit test goodness

File: client/extensions/woocommerce/state/sites/settings/stripe-connect-account/handlers.js [hunk 1]
```diff
@@ -55,6 +64,33 @@ export function handleAccountCreateFailure( { dispatch }, action, error ) {
...
+export function handleAccountDisconnect( { dispatch }, action ) {
+	const { siteId } = action;
+	dispatch(
+		request( siteId, action, '/wc/v1' ).post( 'connect/stripe/account/deauthorize', {
+			deauthorize: true,
+		} )
+	);
+}
+
+export function handleAccountDisconnectSuccess( store, action ) {
```
Comment (budzanowski) [inline client/extensions/woocommerce/state/sites/settings/stripe-connect-account/handlers.js]: destructure dispatch as in above and below.

File: client/extensions/woocommerce/state/sites/settings/stripe-connect-account/selectors.js [hunk 2]
```diff
@@ -22,11 +22,15 @@ const getRawSettings = ( state, siteId ) => {
 	] );
 };
 
+export function getIsDisconnecting( state, siteId = getSelectedSiteId( state ) ) {
```
Comment (budzanowski) [inline client/extensions/woocommerce/state/sites/settings/stripe-connect-account/selectors.js]: maybe documentation

File: client/extensions/woocommerce/state/sites/settings/stripe-connect-account/reducer.js [hunk 3]
```diff
@@ -85,15 +90,52 @@ function connectAccountFetchComplete( state = {}, action ) {
...
 	} );
 }
 
+/**
+ * Updates state to indicate account disconnection request is in progress
+ *
+ * @param  {Object} state  Current state
+ * @return {Object}        Updated state
+ */
+function connectAccountDisconnect( state = {} ) {
```
Comment (budzanowski) [inline client/extensions/woocommerce/state/sites/settings/stripe-connect-account/reducer.js]: I know that this is specific to the product ( stripe connect ) but connectAccountDisconnect sounds confusing. I don't know if there is any better name here - just an observation.

File: client/extensions/woocommerce/state/sites/settings/stripe-connect-account/test/handlers.js [hunk 4]
```diff
@@ -106,6 +110,77 @@ describe( 'handlers', () => {
 		} );
 	} );
 
+	describe( '#handleAccountDisconnect', () => {
```
Comment (budzanowski) [inline client/extensions/woocommerce/state/sites/settings/stripe-connect-account/test/handlers.js]: 👍

File: client/extensions/woocommerce/state/sites/settings/stripe-connect-account/test/reducer.js [hunk 5]
```diff
@@ -343,4 +347,176 @@ describe( 'reducer', () => {
...
+				type: WOOCOMMERCE_SETTINGS_STRIPE_CONNECT_ACCOUNT_DISCONNECT_COMPLETE,
+				siteId: 123,
+				error: 'My error',
+			};
+			const newState = sitesReducer(
+				{
+					123: {
+						settings: {
+							stripeConnectAccount: {
+								connectedUserID: 'acct_14qyt6Alijdnw0EA',
```
Comment (budzanowski) [inline client/extensions/woocommerce/state/sites/settings/stripe-connect-account/test/reducer.js]: same {} as below - maybe extract

**Comment by budzanowski** [comment]
Small nits - nothing blocking.
Complicated stuff.

File: client/extensions/woocommerce/app/settings/payments/stripe/payment-method-stripe-connected-dialog.js [hunk 6]
```diff
@@ -111,22 +112,31 @@ class PaymentMethodStripeConnectedDialog extends Component {
...
 
 		const buttons = [];
 
+		const disabled = isDisconnecting;
+
 		if ( stripeConnectAccount.isActivated ) {
-			buttons.push( { action: 'cancel', label: translate( 'Cancel' ), onClick: onCancel } );
+			buttons.push( {
+				action: 'cancel',
+				disabled: disabled,
```
Comment (justinshreve) [inline client/extensions/woocommerce/app/settings/payments/stripe/payment-method-stripe-connected-dialog.js]: Optional: You can just pass `disabled,` here and in the other objects.

File: client/extensions/woocommerce/state/sites/settings/stripe-connect-account/test/handlers.js [hunk 7]
```diff
@@ -106,6 +110,77 @@ describe( 'handlers', () => {
...
+							message: '.',
+						},
+						success: false,
+					},
+					status: 400,
+				},
+			};
+
+			const action = disconnectAccount( siteId );
+			handleAccountDisconnectFailure( store, action, response.data.body.data.message );
```
Comment (justinshreve) [inline client/extensions/woocommerce/state/sites/settings/stripe-connect-account/test/handlers.js]: Minor: It looks like this is the structure that would be used if you were making a call with http://v2.wp-api.org/reference/global.html#envelope -- While this is still a valid test for the behavior, you could probably reduce the example here a bit and just pass a string.

File: client/extensions/woocommerce/state/sites/settings/stripe-connect-account/test/reducer.js [hunk 8]
```diff
@@ -343,4 +347,176 @@ describe( 'reducer', () => {
 			expect( newState[ 456 ].settings.stripeConnectAccount.isRequesting ).to.eql( true );
 		} );
 	} );
+
+	// TODO
```
Comment (justinshreve) [inline client/extensions/woocommerce/state/sites/settings/stripe-connect-account/test/reducer.js]: Still need these `TODO`s?  Tests look OK to me.

**Comment by allendav** [comment]
FYI everyone - I made quite a few changes in #19115 which will get picked up here as soon as I rebase and edit.

@justinshreve wrote:

> All tests pass. Tests as described. My only question is -- do we want to drop them back on the API key flow? Wouldn't it be better to take them back to the empty connection screen?:

I was wondering that too. @kellychoffman - I am thinking that if they just disconnected @justinshreve is right - let's put them back to the "let's get you connected screen" - they can always say "I want to enter my keys" themselves at that point if need be

**Comment by allendav** [comment]
@justinshreve wrote:

> All tests pass. Tests as described. My only question is -- do we want to drop them back on the API key flow? Wouldn't it be better to take them back to the empty connection screen?:

@allendav wrote:

> I was wondering that too. @kellychoffman - I am thinking that if they just disconnected @justinshreve is right - let's put them back to the "let's get you connected screen" - they can always say "I want to enter my keys" themselves at that point if need be

@allendav then wrote:

On further thought, let's tackle this in a fresh PR - I am sure there will be other transitions we'll want to consider holistically. Let's get all the pieces in an see what shakes out. OK?

File: client/extensions/woocommerce/state/sites/settings/stripe-connect-account/test/reducer.js (hunk 8 above)
Comment (allendav) [reply client/extensions/woocommerce/state/sites/settings/stripe-connect-account/test/reducer.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/19244/commits/2d81a7a15a8b02918fd6b2ef39b9a54500e0b52c

File: client/extensions/woocommerce/state/sites/settings/stripe-connect-account/test/handlers.js (hunk 7 above)
Comment (allendav) [reply client/extensions/woocommerce/state/sites/settings/stripe-connect-account/test/handlers.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/19244/commits/2d81a7a15a8b02918fd6b2ef39b9a54500e0b52c (big rewrite of tests)

File: client/extensions/woocommerce/state/sites/settings/stripe-connect-account/test/reducer.js (hunk 5 above)
Comment (allendav) [reply client/extensions/woocommerce/state/sites/settings/stripe-connect-account/test/reducer.js]: I'd like to leave it as is for now and consider creating some common fixtures after all the flows are hooked up.

File: client/extensions/woocommerce/state/sites/settings/stripe-connect-account/reducer.js (hunk 3 above)
Comment (allendav) [reply client/extensions/woocommerce/state/sites/settings/stripe-connect-account/reducer.js]: Agreed. Renamed everything I could to deauthorize in https://github.com/Automattic/wp-calypso/pull/19244/commits/2d81a7a15a8b02918fd6b2ef39b9a54500e0b52c

File: client/extensions/woocommerce/state/sites/settings/stripe-connect-account/selectors.js (hunk 2 above)
Comment (allendav) [reply client/extensions/woocommerce/state/sites/settings/stripe-connect-account/selectors.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/19244/commits/2d81a7a15a8b02918fd6b2ef39b9a54500e0b52c

File: client/extensions/woocommerce/state/sites/settings/stripe-connect-account/handlers.js (hunk 1 above)
Comment (allendav) [reply client/extensions/woocommerce/state/sites/settings/stripe-connect-account/handlers.js]: Big rewrite in https://github.com/Automattic/wp-calypso/pull/19244/commits/2d81a7a15a8b02918fd6b2ef39b9a54500e0b52c got rid of this

File: client/extensions/woocommerce/app/settings/payments/stripe/payment-method-stripe-connected-dialog.js (hunk 6 above)
Comment (allendav) [reply client/extensions/woocommerce/app/settings/payments/stripe/payment-method-stripe-connected-dialog.js]: Agreed - fixed in https://github.com/Automattic/wp-calypso/pull/19244/commits/ee1e0fee72ef06b3b54d1b39dab0661a7bbf4f54

**Comment by justinshreve** [comment]
> Let's tackle this in a fresh PR

Agreed. We didn't need to block merging this or even launching anything, since it is something we can change independently.

**Comment by kellychoffman** [comment]
> Wouldn't it be better to take them back to the empty connection screen?

Yep! Agree.

**Pull request merged**